aten/src/ATen/BatchedFallback.cpp

```diff
@@ -34,6 +34,172 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
       [] (const Argument& arg) { return arg.type()->isSubtypeOf(ListType::ofTensors()); });
 }
 
+// Returns if an operator is in-place. An operator is inplace if:
+// 1. its name ends in '_'
+// 2. The first argument is a Tensor and it is being written to
+// 3. The first argument is being returned
+// 4. No other arguments are aliased
+// Here is an example of an in-place operator:
+// add_(Tensor(a!) self, Tensor other, *, Scalar alpha=1) -> Tensor(a!)
+static bool isInplaceOp(const c10::FunctionSchema& schema) {
+  if (!schema.is_mutable() || schema.name().back() != '_') {
+    return false;
+  }
+  if (schema.returns().size() != 1) {
+    return false;
+  }
+  // Check that the first argument is being written to
+  const auto& first_arg_alias_info = schema.arguments().begin()->alias_info();
+  if (!first_arg_alias_info || !first_arg_alias_info.value().isWrite()) {
+    return false;
+  }
+  // Check that none of the other args are being aliased
+  for (auto it = schema.arguments().begin() + 1; it != schema.arguments().end(); ++it) {
+    const auto& alias_info = it->alias_info();
+    if (alias_info) {
+      return false;
+    }
+  }
+  // Check that the first tensor is being returned (i.e., output has a (a!))
+  const auto& return_alias_info = schema.returns()[0].alias_info();
+  return return_alias_info && return_alias_info.value().isWrite();
+}
+
+static int64_t firstSetBitPosition(std::bitset<kVmapMaxTensorDims> bdims) {
+  auto bdims_as_int64 = bdims.to_ulong();
+  return std::log2(bdims_as_int64 & -bdims_as_int64) + 1;
+}
+
+static void warnFallback(const c10::FunctionSchema& schema, bool is_inplace) {
+  auto uses_stack = is_inplace ? "" : " and stack";
+  TORCH_WARN("Batching rule not implemented for ", schema.operator_name(), " falling back "
+             "to slow (for loop", uses_stack, ") implementation");
+}
+
+// The general flow of the algorithm is as follows.
+// - First, we figure out which arguments are BatchedTensors and save them
+//   to a vector. We also store a vector of which index of the arguments list
+//   each BatchedTensor appears in. This will be useful for bookkeeping later.
```

Comment:
Wondering if you actually need to extract out the BatchedTensors here or not; maybe a uniform representation which abstracts over batched tensors and non-batched tensors would make the algorithm here more uniform.

Reply:
A uniform representation would save us from storing a vector of indices each BatchedTensor appears in. We could do the following:
- save all Tensor arguments to a vector
- vmap transform all of the BatchedTensors in the Tensor arguments
    - Right now, vmap transforms assume that their inputs are BatchedTensors. We'd have to make the vmap transform API more uniform.
- While iterating through all of the arguments, insert transformed tensors instead of the original tensors onto the stack to be operated on.

I guess we already have a uniform representation - a Tensor can either be batched or not-batched. To make the uniform representation work out we'd have to go and refactor vmap transforms to work with them

Reply:
Because a non-batched tensor is "simply" a batch tensor with no vmap dims.

Reply:
Yup, exactly. I think a refactor of the vmap transforms API is in store; I've been trying to figure out what it should look like